hydra-node/test/Hydra/JSONSchema.hs

```diff
@@ -23,31 +23,79 @@ import System.Exit (ExitCode (..))
 import System.FilePath (normalise, takeBaseName, takeExtension, (<.>), (</>))
 import System.IO.Error (IOError, ioeGetErrorType)
 import System.Process (readProcessWithExitCode)
-import Test.Hydra.Prelude (createSystemTempDirectory, failure)
-import Test.QuickCheck (Property, counterexample, forAllBlind, forAllShrink, vectorOf)
+import Test.Hydra.Prelude (failure, withTempDir)
+import Test.QuickCheck (Property, counterexample, forAllBlind, forAllShrink, resize, vectorOf)
 import Test.QuickCheck.Monadic (assert, monadicIO, monitor, run)
 import qualified Prelude
 
+-- | Validate an 'Arbitrary' value against a JSON schema.
+--
+-- The second argument is a lens that says which part of the JSON file to use to
+-- do the validation, for example:
+--
+-- @@
+-- key "components" . key "schemas" . key "Address"
+-- @@
+--
+-- which selects the JSON schema for "Address" types in a bigger specification,
+-- say an asyncapi description.
+prop_validateJSONSchema ::
+  forall a.
+  (ToJSON a, Arbitrary a, Show a) =>
+  -- | Path to the JSON file holding the schema.
+  String ->
+  -- | Selector into the JSON file pointing to the schema to be validated.
+  SpecificationSelector ->
+  Property
+prop_validateJSONSchema specFileName selector =
+  forAllShrink (resize 10 arbitrary) shrink $ \(samples :: [a]) ->
+    monadicIO $ do
+      withJsonSpecifications $ \tmpDir -> do
+        run ensureSystemRequirements
+        let jsonInput = tmpDir </> "jsonInput"
+        let jsonSchema = tmpDir </> "jsonSchema"
+        let specJsonFile = tmpDir </> (specFileName <> ".json")
+        mSpecs <- run $ Aeson.decodeFileStrict specJsonFile
+        case mSpecs of
+          Nothing -> error "Failed to decode specFile to JSON"
+          Just specs -> run $ do
+            let jsonSpecSchema =
+                  Aeson.object
+                    [ "$id" .= ("file://" <> tmpDir <> "/")
+                    , "type" .= Aeson.String "array"
+                    , "items" .= (specs ^? selector)
+                    ]
+            writeFileLBS jsonInput (Aeson.encode samples)
+            writeFileLBS jsonSchema (Aeson.encode jsonSpecSchema)
+        monitor $ counterexample (decodeUtf8 . Aeson.encode $ samples)
+        (exitCode, _out, err) <- run $ do
+          readProcessWithExitCode "jsonschema" ["-i", jsonInput, jsonSchema] mempty
+        monitor $ counterexample err
+        assert (exitCode == ExitSuccess)
+
 -- | Generate arbitrary serializable (JSON) value, and check their validity
 -- against a known JSON schema.
 -- This property ensures that JSON instances we produce abide by
 -- the specification. Note this, because this uses an external tool each
 -- property iteration is pretty slow. So instead, we run the property only
 -- once, but on a list of 100 elements all arbitrarily generated.
+-- TODO: This should be replacable by 'prop_validateJSONSchema'
```

Comment:
Should: resolve TODO (or explain why not as per PR checklist)

Reply:
Done in https://github.com/input-output-hk/hydra/commit/a9324ebadc482520d6acc3cbc8eb9d11dca70a02